options_test.go

```diff
@@ -18,67 +18,397 @@ package edgedb
 
 import (
 	"errors"
+	"os"
+	"path"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
-func TestParseHost(t *testing.T) {
-	opts, err := DSN("edgedb://me@localhost:5656/somedb")
-	require.Nil(t, err)
-	assert.Equal(t, "localhost", opts.Host)
-}
+func setenvmap(m map[string]string) func() {
+	funcs := make([]func(), 0, len(m))
+	for key, val := range m {
+		funcs = append(funcs, setenv(key, val))
+	}
 
-func TestParsePort(t *testing.T) {
-	opts, err := DSN("edgedb://me@localhost:5656/somedb")
-	require.Nil(t, err)
-	assert.Equal(t, 5656, opts.Port)
+	return func() {
+		for _, fn := range funcs {
+			fn()
+		}
+	}
 }
 
-func TestParseUser(t *testing.T) {
-	opts, err := DSN("edgedb://me@localhost:5656/somedb")
-	require.Nil(t, err)
-	assert.Equal(t, "me", opts.User)
-}
+func setenv(key, val string) func() {
+	old, ok := os.LookupEnv(key)
 
-func TestParseDatabase(t *testing.T) {
-	opts, err := DSN("edgedb://me@localhost:5656/somedb")
-	require.Nil(t, err)
-	assert.Equal(t, "somedb", opts.Database)
-}
+	err := os.Setenv(key, val)
+	if err != nil {
+		panic(err)
+	}
 
-func TestParsePassword(t *testing.T) {
-	opts, err := DSN("edgedb://me:secret@localhost:5656/somedb")
-	require.Nil(t, err)
-	assert.Equal(t, "secret", opts.Password)
-}
+	if ok {
+		return func() {
+			err = os.Setenv(key, old)
+			if err != nil {
+				panic(err)
+			}
+		}
+	}
 
-func TestMissingPort(t *testing.T) {
-	opts, err := DSN("edgedb://me@localhost/somedb")
-	require.Nil(t, err)
-	assert.Equal(t, 5656, opts.Port)
+	return func() {
+		err = os.Unsetenv(key)
+		if err != nil {
+			panic(err)
+		}
+	}
 }
 
-func TestDialHost(t *testing.T) {
-	opts := Options{Host: "some.com", Port: 1234}
-	assert.Equal(t, "some.com:1234", opts.address())
+func TestConUtils(t *testing.T) {
```

Comment:
Did you copy these from edgedb-python?

Reply:
This was an attempt to copy what was in edgedb-python. I went back and found that I had several errors. I also compared this with edgedb-js to be sure there wasn't anything different there.

Reply:
Yeah, that's awesome. I was only asking because I *thought* the implementation there was good tested :)